### PR TITLE
PageView scroll physics to match Android (as an opt-in)

### LINKED ISSUE #38357: TabView swipe animates too slow compared to Android
## Use case

When you swipe through your tabs, it takes quite some time until the animation is finished. So you really have to wait until you can scroll down. It also slows down even more at the end of the animation. On native android apps this animation is way faster...

## Proposal

Please make the animation similar to android.

Thanks in advance :)

### PR DESCRIPTION
**Pull request by nt4f04uNd**
Fixes https://github.com/flutter/flutter/issues/38357

So, this is a new reland for a https://github.com/flutter/flutter/pull/95423
Previous reland https://github.com/flutter/flutter/pull/97166

<details>
<summary>Old physics</summary>

https://user-images.githubusercontent.com/39104740/146440807-50e10ab1-628c-4623-a9d3-3704bd5fd0d5.mp4

https://user-images.githubusercontent.com/39104740/205435343-57a338b8-f04c-4d6d-bec5-b677dbb6e1eb.mov

</details>

<details>
<summary>New physics</summary>

https://user-images.githubusercontent.com/39104740/146440570-ba140add-0411-454a-9456-34d6d62e9fc8.mp4

https://user-images.githubusercontent.com/39104740/205434169-8a7197cd-8c00-4d21-8f56-33cf6d58f838.mov

</details>

### Opt-in

This change has been proven to be breaking for the customer tests.
To opt in, users should provide `pageSnappingPhysics: const PageScrollPhysics(useNewPhysics: true)`

If tests are broken, this is likely because of the change in `minFlingVelocity` or `minFlingDistance`.
Previous values are respectively `50.0` and `18.0`.
New values are `400.0` and `25.0`.

To migrate the tests, one should change the drag velocity/distance to match the new minimum values.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by nt4f04uNd** [comment]
Adding `pageSnappingPhysics` also allows for customization of the simulation (and in particular, to use the old simulation).
This might be a sufficient fix for https://github.com/flutter/flutter/issues/55103

**Comment by nt4f04uNd** [comment]
Is https://github.com/flutter/flutter/pull/97166#discussion_r874089028 really blocking us from landing this PR? Seems pretty harmless

**Comment by nt4f04uNd** [comment]
It isn't a severe API break, it only breaks some tests, so should I really go through all the process outlined in the breaking change policy? https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

This seems a bit excessive, given that there's nothing to actually migrate here, after an opt-in is removed.

**Comment by Piinks** [comment]
> after an opt-in is removed.

How will folks know to migrate without guidance? :)

There is a lot here I'd like to look at closely, I don't recall the specific failures that caused it to be reverted before. I'll be able to take a close look at this tomorrow. Nice to see you back!

File: packages/flutter/test/widgets/scroll_physics_test.dart [hunk 1]
```diff
@@ -111,13 +111,25 @@ void main() {
...
+    expect(clamp.createBallisticSimulation(position, 1000)!.dx(0), 1000);
+
+    // Contrary to the other two scroll physics, PageScrollPhysics do intentionally produce
+    // a bigger velocity than supplied to them, but they won't suffer from this, because the simulation
+    // velocity depends on a ratio of velocity / pageDelta, and because they are not affected by
+    // https://github.com/flutter/flutter/issues/11599
+
+    // TODO(nt4f04uNd): remove this test, because it is artificial, in favor of a real one.
+    // Scroll simulations surely should be allowed to modify the speed they receive in whatever way they want.
+    // The proper fix for the original bug https://github.com/flutter/flutter/issues/24715
```
Comment (Piinks) [inline packages/flutter/test/widgets/scroll_physics_test.dart:129]: This issue is closed, is it not actually fixed based on this comment?

File: packages/flutter/test/widgets/scroll_physics_test.dart [hunk 2]
```diff
@@ -111,13 +111,25 @@ void main() {
...
     const ClampingScrollPhysics clamp = ClampingScrollPhysics();
-    const PageScrollPhysics page = PageScrollPhysics();
 
+    // Verify creating these simulations doesn't alter the velocity for time 0.
+    //
     // Calls to createBallisticSimulation may happen on every frame (i.e. when the maxScrollExtent changes)
-    // Changing velocity for time 0 may cause a sudden, unwanted damping/speedup effect
-    expect(bounce.createBallisticSimulation(position, 1000)!.dx(0), moreOrLessEquals(1000));
-    expect(clamp.createBallisticSimulation(position, 1000)!.dx(0), moreOrLessEquals(1000));
-    expect(page.createBallisticSimulation(position, 1000)!.dx(0), moreOrLessEquals(1000));
```
Comment (Piinks) [inline packages/flutter/test/widgets/scroll_physics_test.dart:120]: I still do not understand these changes, why are bouncing and clamping scroll physics affected by this change?
Why can we not validate the page physics here anymore? What happens in this case with this change in page physics? What will the experience be for a user if this ballistic simulation is created?

File: packages/flutter/test/widgets/scroll_physics_test.dart [hunk 3]
```diff
@@ -111,13 +111,25 @@ void main() {
...
+    // Changing velocity for time 0 may cause a sudden, unwanted damping/speedup effect.
+    expect(bounce.createBallisticSimulation(position, 1000)!.dx(0), 1000);
+    expect(clamp.createBallisticSimulation(position, 1000)!.dx(0), 1000);
+
+    // Contrary to the other two scroll physics, PageScrollPhysics do intentionally produce
+    // a bigger velocity than supplied to them, but they won't suffer from this, because the simulation
+    // velocity depends on a ratio of velocity / pageDelta, and because they are not affected by
+    // https://github.com/flutter/flutter/issues/11599
+
+    // TODO(nt4f04uNd): remove this test, because it is artificial, in favor of a real one.
```
Comment (Piinks) [inline packages/flutter/test/widgets/scroll_physics_test.dart:127]: I don't follow this TOD, what is the action item or issue that needs to be revisited here?

File: packages/flutter/lib/src/material/tabs.dart [hunk 4]
```diff
@@ -1420,6 +1420,7 @@ class TabBarView extends StatefulWidget {
     required this.children,
     this.controller,
     this.physics,
+    this.pageSnappingPhysics,
```
Comment (Piinks) [inline packages/flutter/lib/src/material/tabs.dart:1423]: What happens if someone provides `physics`?

File: packages/flutter/lib/src/material/tabs.dart [hunk 5]
```diff
@@ -1444,11 +1445,19 @@ class TabBarView extends StatefulWidget {
...
   /// The physics are modified to snap to page boundaries using
-  /// [PageScrollPhysics] prior to being used.
+  /// [pageSnappingPhysics] prior to being used.
   ///
   /// Defaults to matching platform conventions.
   final ScrollPhysics? physics;
 
+  /// Controls how pages should snap to view boundaries.
+  ///
+  /// Prior being used, ombined with other scroll physics
```
Comment (Piinks) [inline packages/flutter/lib/src/material/tabs.dart:1455]: Isn't this something the user could provide to `physics`?

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 6]
```diff
@@ -532,13 +533,36 @@ class _ForceImplicitScrollPhysics extends ScrollPhysics {
...
 class PageScrollPhysics extends ScrollPhysics {
   /// Creates physics for a [PageView].
-  const PageScrollPhysics({ super.parent });
+  const PageScrollPhysics({ super.parent, this.useNewPhysics = false });
+
+  /// An opt-in flag for a new implementation of physics.
+  ///
+  /// Enabling this will use scroll physics which are more close to native
+  /// Android and iOS implementations.
+  final bool useNewPhysics;
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/page_view.dart:546]: How should customers migrate? Do they need to provide a custom scroll physics to every Tab or Page view? How do we migrate them off of it?

File: packages/flutter/lib/src/widgets/page_view.dart [hunk 7]
```diff
@@ -532,13 +533,36 @@ class _ForceImplicitScrollPhysics extends ScrollPhysics {
 ///  * [ScrollPhysics], the base class which defines the API for scrolling
 ///    physics.
 ///  * [PageView.physics], which can override the physics used by a page view.
+///  * [PageScrollSimulation], which implements Android page view scroll physics, and
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/page_view.dart:536]: Flutter does not just run on Android. Does this adapt to the platform?

**Review by Piinks — COMMENTED** [review]
I refreshed my memory on this. At the time, the customers affected looked like simple fixes, but reading through this I am not really sure how to migrate them. Can you advise?

**Comment by Piinks** [comment]
> I'm sorry this is taking a long time for me to respond, but I do remember about this PR and will update it as soon as I can.

No worries at all @nt4f04uNd. I am going to close this for now so it is off our review queue. When you are ready to work on it again, please feel free to reopen it. I'll be here ready to review it. :)

**Comment by Liloupar** [comment]
@nt4f04uNd 
Do you still have time to take another look at this PR?

**Pull request closed without merging**